File: data-structures/tree-traversals/visual-basic/bst_traverse_pre_order.vb
' Raspberry Pi Foundation
' Developed as part of Ada Computer Science
' Usage licensed under CC BY-NC-SA 4.0
' Note: This file is designed to be copied out and compiled on your machine.
' To run this file you need to:
' 1. Copy the contents
' 2. Paste them into the Visual Basic IDE of your choice (Visual Studio, for example)
' 3. Compile the program
' 4. Run the program


' A class to represent a node in a binary search tree
Class Node
    Public data As Integer ' Tree is storing integer values
    Public left As Node
    Public right As Node

    ' Constructor method
    Public Sub New(ByVal itemData As Integer)
        data = itemData
    End Sub

End Class


' A class to represent a binary search tree (BST)
Class BST
    Private root As Node ' Do not intialise the root yet as the tree is empty

    ' Return the root node
    Public Function GetRoot() As Node
        Return root
    End Function

    ' Insert a new node
    Public Sub Insert(ByVal item As Integer)
        ' Create a new node
        Dim newNode As Node = New Node(item)

        ' Check if the root node exists
        If root Is Nothing Then
            ' Set the root to be the new node
            ' Find the correct place to insert the new node
            root = newNode
        Else
            Dim current = root
            Dim placed = False

            ' Repeat while the data has not been inserted
            While placed = False
                ' Check if the new item is greater than the current node data
                If item > current.data Then
                    ' Check if the current node does not have a right child node
                    If current.right Is Nothing Then
                        ' Insert the new node to the right of the current node
                        current.right = newNode
                        ' Otherwise repeat with the current right node
                        placed = True
                    Else
                        current = current.right

                        ' Otherwise the new item is less than or equal to the current node
                    End If
                Else
                    ' Check if the current node does not have a left child node
                    If current.left Is Nothing Then
                        ' Insert the new node to the left of the current node
                        current.left = newNode
                        ' Otherwise repeat with the current left node
                        placed = True
                    Else
                        current = current.left
                    End If
                End If
            End While
        End If
    End Sub


    ' A Pre-order traversal of the binary search tree
    Public Sub PreOrderTraversal(ByVal node As Node)
        ' Output the data of the current node
        Console.WriteLine(node.data)

        ' Check any nodes to the left of the current node
        If node.left IsNot Nothing Then
            PreOrderTraversal(node.left)
        End If

        ' Check any nodes to the right of the current node
        If node.right IsNot Nothing Then
            PreOrderTraversal(node.right)
        End If
    End Sub


End Class


Module program

    ' The Main method is the entry point for all C# programs
    Sub Main()
        ' Instantiate a new tree object
        Dim bst As BST = New BST()

        ' Insert test data into the binary search tree
        InsertTestData(bst)

        ' Get the root node
        Dim root As Node = bst.GetRoot()

        ' Testing
        Console.WriteLine("### Output the tree structure ###")
        Console.WriteLine("The root is the left-most value and the children are to the right" & vbLf)
        OutputTree(root)

        ' Output the binary search tree using an pre-order traversal
        Console.WriteLine(vbLf & "### Binary search tree (pre-order traversal) ###")
        bst.PreOrderTraversal(root)
    End Sub


    ' Insert test data into the binary search tree
    Sub InsertTestData(ByVal bst As BST)
        bst.Insert(8)  ' This will be the root node
        bst.Insert(3)
        bst.Insert(6)
        bst.Insert(10)
        bst.Insert(1)
        bst.Insert(5)
        bst.Insert(18)
        bst.Insert(7)
        bst.Insert(12)
    End Sub


    ' Output the tree with the root to the left and children to the right
    Sub OutputTree(ByVal node As Node, ByVal Optional level As Integer = 0)
        If node IsNot Nothing Then
            OutputTree(node.right, level + 1)

            Dim spaces As String = (" " * 4 * level) ' Variable length string of spaces
            Console.WriteLine($"{spaces}-> {node.data}")

            OutputTree(node.left, level + 1)
        End If
    End Sub


End Module

File: data-structures/tree-traversals/python/bst_traverse_pre_order.py
# Raspberry Pi Foundation
# Developed as part of Ada Computer Science
# 
# Usage licensed under CC BY-NC-SA 4.0


class Node:
    """A node in a binary search tree"""
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

    def get_data(self):
        return self.data

    def get_left(self):
        return self.left

    def set_left(self, new_left):
        self.left = new_left

    def get_right(self):
        return self.right

    def set_right(self, new_right):
        self.right = new_right


class BST:
    """A class for constructing a binary search tree (BST)"""
    
    def __init__(self):
        """Constructor method"""
        self.root = None  # Initialise to None as the tree is empty
    
    def get_root(self):
        """Return the root node"""
        return self.root

  
    def insert(self, item):
        """Insert a new node"""

        # Create a new node
        new_node = Node(item)

        # Check if the root node exists
        if self.root is None:
            # Set the root to be the new node
            self.root = new_node

        # Find the correct place to insert the new node
        else:
            current = self.root
            placed = False

            # Repeat while the data has not been inserted            
            while placed == False:
                
                # Check if the new item is greater than the current node data
                if item > current.get_data():
                    # Check if the current node does not have a right child node
                    if current.get_right() is None:
                        # Insert the new node to the right of the current node
                        current.set_right(new_node)
                        placed = True
                    # Otherwise repeat with the current right node
                    else:
                        current = current.get_right()

                # Otherwise the new item is less than or equal to the current node
                else:
                    # Check if the current node does not have a left child node
                    if current.get_left() is None:
                        # Insert the new node to the left of the current node
                        current.set_left(new_node)
                        placed = True
                    # Otherwise repeat with the current left node
                    else:
                        current = current.get_left()          

    
    def pre_order_traversal(self, node):
        """A pre-order traversal of the binary search tree"""

        # Output the data of the current node
        print(node.get_data())

        # Check any nodes to the left of the current node
        if node.get_left() is not None:
            self.pre_order_traversal(node.get_left())
        
        # Check any nodes to the right of the current node
        if node.get_right() is not None:
            self.pre_order_traversal(node.get_right())
            

def insert_test_data(bst):
    """"Insert test data into the binary search tree"""
    bst.insert(8)  # This will be the root node
    bst.insert(3)
    bst.insert(6)
    bst.insert(10)
    bst.insert(1)
    bst.insert(5)
    bst.insert(18)
    bst.insert(7)
    bst.insert(12)


def output_tree(node, level = 0):
    """Print the tree with the root to the left and children to the right"""

    if node is not None:
        output_tree(node.get_right(), level + 1)
        print(' ' * 4 * level + '-> ' + str(node.get_data()))
        output_tree(node.get_left(), level + 1)


def main():
    """Create a binary search tree and search for an item"""

    # Instantiate a new tree object
    bst = BST()

    # Insert test data into the binary search tree
    insert_test_data(bst)

    # Get the root node
    root = bst.get_root()

    # Testing
    print("### Output the tree structure ###")
    print("The root is the left-most value and the children are to the right\n")
    output_tree(root)

    # Output the binary search tree using a pre-order traversal
    print("\n### Binary search tree (pre-order traversal) ###")
    bst.pre_order_traversal(root)
    

# This code will run if this file is executed directly
# (i.e. not called by another program)
if __name__ == "__main__":
    main()
